Add test for fixing mismatched versions in lockfile

rwstauner opened the pull request:
This test confirms that a bug that existed in 2.6.3
was fixed by b8e55087f026ffcc5ebf135ca6726daf0ec40b5d

We had an issue where we were pointing to a git repo
and a bad rebase (or something) occurred and replaced platform specific lines of a different (higher) version.
Bundler (2.6.2 at the time) chose the platform specific gem over the git gem because of it.

Thankfully I discovered that this has been fixed, as long as the git repo locks the version, which isn't something I had previously thought to do with a git gem.

So first question: Is this test worth merging?
Second question: Is it expected/desired that if a git gem does not specify a version, that platform lines in the lock file would take precedence?  If that sounds confusing and we don't want that, this test can verify that behavior by just removing the version lock (and then it will fail).

deivid-rodriguez commented:
Hei!

I went to confirm that b8e55087f026ffcc5ebf135ca6726daf0ec40b5d "fixed" this, but I found it was actually a side-effect of d8c4754d8f3952b79228d43cc9ef930f3763a0b3.

> So first question: Is this test worth merging?

I think it's fine to merge, sure. However, can you remove `focus: true`. Also, can you reword the spec to something like "automatically fixes the lockfile when it includes a gem under the correct GIT section, but also under an incorrect GEM section, with a higher version".

> Is it expected/desired that if a git gem does not specify a version, that platform lines in the lock file would take precedence?

That sounds incorrect to me. What happens if you try to use that lockfile? Hopefully the lockfile version will still be ignored, but still Bundler could do better and correct it.

rwstauner commented:
Sorry about that, I removed the `focus`.
I updated the spec title as you suggested 👍 

As for
> That sounds incorrect to me. What happens if you try to use that lockfile? Hopefully the lockfile version will still be ignored, but still Bundler could do better and correct it.

Unfortunately no, what we found is that it would take the platform-specific (newer) version and ignore the git repo.
It's possible that this depends on what is already installed/cached, I didn't test it extensively, but we had issues with the wrong version being selected.

deivid-rodriguez commented:
I don't understand how Bundler would end up using the version under the incorrect source but in any case, I proposed a fix to correct the incorrect lockfile at #8579. I included your additional test in that PR, so let me close this in favor of that other PR 👍.
